tutorials-v5/miscellaneous/v5_paper-optimal-control.md

```diff
@@ -0,0 +1,299 @@
+Whereas the former two have been part of the `QuTiP-QTRL` package of QuTiPv4, the latter is a new addition in version 5.
+Allthogether, these algorithms are now included in the new `QuTiP-QOC` package that also adds additional functionalities such as integration with JAX via the JAX optimization technique (JOPT).
+
+```python
+import matplotlib.pyplot as plt
+import numpy as np
+from jax import jit, numpy
+from qutip import (about, gates, liouvillian, qeye, sigmam, sigmax, sigmay,
+                   sigmaz)
+from qutip_qoc import Objective, optimize_pulses
+
+%matplotlib inline
+```
+
+## Introduction
+
+In this example we want to implement a Hadamard gate on a single qubit.
+In general, a qubit might be subject to decoherence which can be captured using the Lindblad formalism with the jump operator $\sigma_{-}$.
+
+For simplicity, we consider a control Hamiltonian parametrized by $\sigma_x$, $\sigma_y$ and $\sigma_z$:
+
+$H_c(t) = c_x(t) \sigma_x + c_y(t) \sigma_y + c_z(t) \sigma_z$
+
+with $c_x(t)$, $c_y(t)$ and $c_z(t)$ as independent control parameters.
+Additionally, we model a constant drift Hamiltonian
+
+$H_d = \dfrac{1}{2} (\omega \sigma_z + \delta \sigma_x)$,
+
+with associated energy splitting $\omega$ and tunneling rate $\delta$.
+The amplitude damping rate for the collapse operator $C = \sqrt{\gamma} \sigma_-$ is denoted as $\gamma$.
+
+```python
+# energy splitting, tunneling, amplitude damping
+omega = 0.1  # energy splitting
+delta = 1.0  # tunneling
+gamma = 0.1  # amplitude damping
+sx, sy, sz = sigmax(), sigmay(), sigmaz()
+
+Hc = [sx, sy, sz]  # control operator
+Hc = [liouvillian(H) for H in Hc]
+
+Hd = 1 / 2 * (omega * sz + delta * sx)  # drift term
+Hd = liouvillian(H=Hd, c_ops=[np.sqrt(gamma) * sigmam()])
+
+# combined operator list
+H = [Hd, Hc[0], Hc[1], Hc[2]]
+```
+
+```python
+# objective for optimization
+initial = qeye(2)
+target = gates.hadamard_transform()
+```
+
+```python
+# pulse time interval
+times = np.linspace(0, np.pi / 2, 100)
+```
+
+## Implementation
+
+### GRAPE Algorithm
+
+The GRAPE algorithm works by minimizing an infidelity loss function that measures how close the final state or unitary tranformation is to the desired target.
+Starting from the provided `guess` control pulse, it optimizes evenly spaced piecewise constant pulse amplitudes.
+In the end, it strives to achieve the desired target infidelity, sepcified by the `fid_err_targ` keyword.
+
+```python
+res_grape = optimize_pulses(
+    objectives=Objective(initial, H, target),
+    control_parameters={
+        "ctrl_x": {"guess": np.sin(times), "bounds": [-1, 1]},
+        "ctrl_y": {"guess": np.cos(times), "bounds": [-1, 1]},
+        "ctrl_z": {"guess": np.tanh(times), "bounds": [-1, 1]},
+    },
+    tlist=times,
+    algorithm_kwargs={"alg": "GRAPE", "fid_err_targ": 0.01},
+)
```

Comment:
Maybe here it would be nice to show the optimization result a bit.

I was trying this and got some weird results.

Reply:
I am mostly confused by the predefined control parameters here for the GRAPE method. 

I think these initial guesses are not used at all in the GRAPE algorithm.

https://github.com/qutip/qutip-qoc/blob/881fb3939a606bdf544a78f4f4ffeb8340af71b2/src/qutip_qoc/objective.py#L98

Reply:
Can you tell me, what you mean with *weird results*? In my case, the plots of the paper are reproduced and then shown at the end of the notebook in the *comparison* section

Reply:
Ok, let me update my comment... after rerunning this, I also get a different result than in the paper. But it seems to be limited to the CRAB algorithm. This is what I get:

![output](https://github.com/user-attachments/assets/b3288d08-077a-4e43-b77e-847830a18689)

Reply:
I think it is not entirely surprising that one gets different results on different machines. The optimization results highly depend on the initial value and optimization methods used. As long as the result fidelity is good, it should be correct.

Could you copy the fidelity you got on your machine? I got a very weird 0.75 instead of 99%

Reply:
This is the shape I got:

![image](https://github.com/user-attachments/assets/6418afaf-e290-483a-859f-58463f7add2f)

Reply:
The first run today also gave me a fidelity around 80% and I reproduced my plot from earlier. But now, after running it about ten times, I always arrive at >99% and get the exact same plot as in the paper. 
I also tried with various initial conditions, but it always reaches the fidelity target.

Reply:
Is it a good idea to add the (in)fidelity check as a test maybe?

Reply:
> The first run today also gave me a fidelity around 80% and I reproduced my plot from earlier. But now, after running it about ten times, I always arrive at >99% and get the exact same plot as in the paper.

This is very confusing, why did running this multiple times change the result? That sounds very dangerous, like some parameters are not correctly reset between each execution? What will happen if one resets the numpy random number generator?

Yes, adding the infidelity check as a test would be very nice.

Reply:
@flowerthrower @ajgpitch any thoughts on this?

Reply:
> This is very confusing, why did running this multiple times change the result? That sounds very dangerous, like some parameters are not correctly reset between each execution? What will happen if one resets the numpy random number generator before each repeat?

After retrying this a couple times today, I always arrive at the result of the paper.

> Yes, adding the infidelity check as a test would be very nice.

Added ✅